std/algorithm/comparison.d

```diff
@@ -892,116 +892,353 @@ nothrow pure @safe unittest
+                {
+                    if (!equal(rs[0], r)) { return false; }
+                }
+                else static if (isAutodecodableString!(typeof(rs[0])) &&
+                                isAutodecodableString!(typeof(r)))
+                {
+                    if (!equal(rs[0].byCodeUnit, r.byCodeUnit)) { return false; }
+                }
+                else static if (isAutodecodableString!(typeof(rs[0])))
+                {
+                    if (!equal(rs[0].byCodeUnit, r)) { return false; }
+                }
+                else static if (isAutodecodableString!(typeof(r)))
+                {
+                    if (!equal(r.byCodeUnit, rs[0])) { return false; }
+                }
             }
+            return true;
         }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        else
         {
-            immutable len1 = r1.length;
-            immutable len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
+            import std.meta : staticMap, staticIndexOf;
+
+            // get lengths
+            alias hasLengthBits = staticMap!(hasLength, Rs);
+            enum indexOfFirstLength = staticIndexOf!(true, hasLengthBits);
+            enum someHasLength = indexOfFirstLength != -1;
 
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            static if (someHasLength && // if one range has a (finite) length and
+                       anySatisfy!(isInfinite, Rs)) // another is infinite
             {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                return false;   // they can't be equal
             }
-            return true;
-        }
-        else
-        {
-            //Generic case, we have to walk both ranges making sure neither is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            else
             {
-                if (r2.empty) return false;
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                enum someHasIndexingAndLength = anySatisfy!(hasIndexingAndLength, Rs);
+
+                static if (someHasLength)
+                {
+                    /* if any `rs` `hasLength` then `primaryRangeIndex` will index the
+                     * first of them */
+                    enum primaryRangeIndex = indexOfFirstLength;
+                    const length = rs[indexOfFirstLength].length;
+                }
+                else
+                {
+                    enum primaryRangeIndex = 0; // just pick first
+                }
+                alias r_ = rs[primaryRangeIndex]; // shorthand
```

Comment:
`r_` is rather non descriptive 20 lines below. How about `firstRange`, `startRange` or `primaryRange`?

Reply:
Done.